## Level 1: The Basics (Initialization & Anatomy)
Goal: Master the creation of arrays and understanding their structure.

Exercise 1.1: The Creator
Task: Create the following arrays without typing the numbers manually (use NumPy functions):

In [2]:
import numpy as np

1. A vector of 10 zeros (integer type).

In [3]:
first_vec=np.zeros(10,dtype=int)
print(first_vec)

[0 0 0 0 0 0 0 0 0 0]


2. A vector of 10 ones (float type).

In [4]:
second_vec = np.ones(10,dtype=float)
print(second_vec)

[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]


3. A vector containing the number 5 repeated 10 times.

In [5]:
third_vec = np.full(10, 5)
print(third_vec)

[5 5 5 5 5 5 5 5 5 5]


4. A vector of integers from 10 to 50 (inclusive).

In [6]:
forth_vec = np.arange(10, 51)
print(forth_vec)

[10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33
 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50]


5. A vector of even integers from 10 to 50 (inclusive).

In [7]:
fifth_vec = np.arange(10, 51, 2)
print(fifth_vec)

[10 12 14 16 18 20 22 24 26 28 30 32 34 36 38 40 42 44 46 48 50]


6. A vector of 5 values evenly spaced between 0 and 1.

In [ ]:
sixth_vec = np.linspace(0, 1, 5)
print(sixth_vec)

[0.   0.25 0.5  0.75 1.  ]


Exercise 1.2: The Inspector
Task: Create a random matrix of shape (3, 4). Write code to print the following details:

In [1]:
matrix = np.random.rand(3, 4)

NameError: name 'np' is not defined

1. The shape of the array.


In [10]:
print("the shape is:",matrix.shape)

the shape is: (3, 4)


2. The total number of elements (size).

In [11]:
print("the size is:",matrix.size)

the size is: 12


3. The number of dimensions (ndim).

In [12]:
print("the Dimensions is:", matrix.ndim)

the Dimensions is: 2


4. The data type of the elements.

In [13]:
print("Data type:", matrix.dtype)

Data type: float64


5. The memory usage in bytes.

In [14]:
print("Memory (bytes):", matrix.nbytes)

Memory (bytes): 96


Exercise 1.3: The Random Generator
Task:

1. Set the random seed to 42 (for reproducibility).

In [ ]:
np.random.seed(42)

2. Generate a 1D array of 10 random integers between 0 and 100.

In [16]:
ints_vec = np.random.randint(0, 101, 10)
print(ints_vec)

[51 92 14 71 60 20 82 86 74 74]


3. Generate a 1D array of 10 random standard normal values (mean=0, std=1).

In [ ]:
normals_vec = np.random.randn(10)

print(normals_vec)

print("------------------------------------------------------")

arr = np.random.normal(loc=0, scale=1, size=10)
print(arr)

[ 0.27904129  1.01051528 -0.58087813 -0.52516981 -0.57138017 -0.92408284
 -2.61254901  0.95036968  0.81644508 -1.523876  ]
------------------------------------------------------
[-0.42804606 -0.74240684 -0.7033438  -2.13962066 -0.62947496  0.59772047
  2.55948803  0.39423302  0.12221917 -0.51543566]


4. Find the maximum and minimum values of your integer array.

In [18]:
print(np.max(arr))
print(np.min(arr))

2.5594880310377928
-2.1396206560762394


## Level 2: Slicing, Indexing & Reshaping
Goal: Learn how to grab specific pieces of data and change dimensions.
Task: Write single-line commands to extract:

In [19]:
arr = np.arange(1, 26).reshape(5, 5)
print(arr)

[[ 1  2  3  4  5]
 [ 6  7  8  9 10]
 [11 12 13 14 15]
 [16 17 18 19 20]
 [21 22 23 24 25]]


1. The row at index 2 ([11, 12, 13, 14, 15]).

In [20]:
print(arr[2])

[11 12 13 14 15]


2. The column at index 3 ([4, 9, 14, 19, 24]).

In [21]:
print(arr[3])

[16 17 18 19 20]


3. The sub-matrix:
    [[12, 13],
    [17, 18]]

In [22]:
print(arr[2:4, 1:3])

[[12 13]
 [17 18]]


5. The last two columns of the last two rows.

In [23]:
print(arr[3:, 3:])

[[19 20]
 [24 25]]


### Exercise 2.2: The Reshaper
Task:

1. Create a 1D array of 12 elements (0 to 11).

In [24]:
arr=np.arange(0,12)
print(arr)

[ 0  1  2  3  4  5  6  7  8  9 10 11]


2. Reshape it into a matrix of 3 rows, 4 columns.


In [25]:
matrix_3_4 = arr.reshape(3, 4)
print(matrix_3_4)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


3. Reshape it into a matrix of 2 rows, 6 columns.

In [26]:
matrix_2_6 = arr.reshape(2, 6)
print(matrix_2_6)

[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]]


4. Try to reshape it into (3, 5). What error do you get? Explain why.
--> ValueError: cannot reshape array of size 12 into shape (3,5)
as the Reshape requires that the total number of elements stays the same.
3 × 5 = 15 elements required, but the array only has 12.

In [27]:
#matrix_3_5 = arr.reshape(3, 5)
#print(matrix_3_5)

#### Exercise 2.3: The Stacker
Setup:
    • A = np.array([[1, 2], [3, 4]])
    • B = np.array([[5, 6], [7, 8]])
Task:


In [28]:
A = np.array([[1, 2], [3, 4]])
B = np.array([[5, 6], [7, 8]])

1. Stack them vertically (Result should be 4x2).

In [29]:
vertical_stack = np.vstack((A, B))
print(vertical_stack)

[[1 2]
 [3 4]
 [5 6]
 [7 8]]


2. Stack them horizontally (Result should be 2x4).

In [30]:
horizontal_stack = np.hstack((A, B))
print(horizontal_stack)

[[1 2 5 6]
 [3 4 7 8]]


### Level 3: Math, Broadcasting & Aggregation
Goal: Perform calculations on whole datasets instantly.
Exercise 3.1: The Broadcaster (Crucial)

In [31]:
matrix = np.ones((3, 3))
vector = np.array([0, 1, 2])
print(matrix)
print(vector)

[[1. 1. 1.]
 [1. 1. 1.]
 [1. 1. 1.]]
[0 1 2]


1. Add the vector to the matrix. Explain the result. Does it add to rows or columns?

In [32]:
result = matrix + vector
print(result)

[[1. 2. 3.]
 [1. 2. 3.]
 [1. 2. 3.]]


2. Reshape the vector to (3, 1) and add it to the matrix. How does the result change?

In [33]:
vector=vector.reshape(3,1)
print(vector)
result = matrix + vector
print(result)


[[0]
 [1]
 [2]]
[[1. 1. 1.]
 [2. 2. 2.]
 [3. 3. 3.]]


Exercise 3.2: The Aggregator
Setup: Generate a random matrix of shape (10, 5) representing 10 students and 5 exams.
Task:

In [34]:
scores = np.random.randint(0, 101, size=(10, 5))
print(scores)

[[20 72 38 17  3]
 [88 59 13  8 89]
 [52  1 83 91 59]
 [70 43  7 46 34]
 [77 80 35 49  3]
 [ 1  5 53  3 53]
 [92 62 17 89 43]
 [33 73 61 99 13]
 [94 47 14 71 77]
 [86 61 39 84 79]]


1. Calculate the average score for each student (Row-wise mean).

In [35]:
student_avg = scores.mean(axis=1)
print("Average score per student:\n", student_avg)

Average score per student:
 [30.  51.4 57.2 40.  48.8 23.  60.6 55.8 60.6 69.8]


2. Calculate the highest score recorded in each exam (Column-wise max).

In [36]:
student_avg = scores.max(axis=0)
print("Average score per student:\n", student_avg)


Average score per student:
 [94 80 83 99 89]


3. Calculate the total variance of the entire dataset.

In [37]:
total_variance = scores.var()
print("Total variance of dataset:", total_variance)

Total variance of dataset: 935.3216


X = np.random.rand(100, 3) * 10  

In [38]:
X = np.random.rand(100, 3) * 10  
print(X)

[[7.47320110e+00 5.39692132e+00 5.86751166e+00]
 [9.65255307e+00 6.07034248e+00 2.75999182e+00]
 [2.96273506e+00 1.65266939e+00 1.56364067e-01]
 [4.23401481e+00 3.94881518e+00 2.93488175e+00]
 [1.40798227e-01 1.98842404e+00 7.11341953e+00]
 [7.90175541e+00 6.05959975e+00 9.26300879e+00]
 [6.51077026e+00 9.14959676e+00 8.50038578e+00]
 [4.49450674e+00 9.54101165e-01 3.70818252e+00]
 [6.68841253e+00 6.65922357e+00 5.91297788e+00]
 [2.74721793e+00 5.61243426e+00 3.82926875e+00]
 [9.71712095e+00 8.48913824e+00 7.21729521e+00]
 [2.35984920e+00 2.56068323e+00 4.04335895e-01]
 [7.10662890e+00 1.10890821e+00 4.39336502e+00]
 [2.01719202e+00 8.95763596e+00 4.75370223e+00]
 [5.63275572e+00 6.95516086e+00 1.39331454e+00]
 [6.04417379e+00 5.39841091e+00 2.03061225e+00]
 [9.42853571e+00 5.98865466e+00 6.94784933e+00]
 [8.80467839e+00 6.24354048e+00 2.95633686e+00]
 [1.05494260e+00 4.56534570e+00 2.18440437e+00]
 [4.16509948e+00 8.83280259e+00 3.24345021e+00]
 [1.22087955e+00 3.56297838e+00 9.068284

1. Compute the mean vector (size 3) and standard deviation vector (size 3).

In [39]:
mean_vector = X.mean(axis=0)
std_vector = X.std(axis=0)
print(mean_vector)
print(std_vector)

[5.02089928 4.95287742 4.95144399]
[2.92397895 2.87587701 2.78602355]


2. Standardize the dataset: Subtract the mean and divide by the standard deviation.

In [40]:
normalization = (X - mean_vector) / std_vector
print(normalization)

[[ 0.83868655  0.15440295  0.3288083 ]
 [ 1.58402433  0.38856497 -0.78658781]
 [-0.7038916  -1.14754839 -1.72111967]
 [-0.26911427 -0.34913254 -0.72381378]
 [-1.66899322 -1.03079978  0.77600763]
 [ 0.985252    0.3848295   1.5475694 ]
 [ 0.50953546  1.45928331  1.27383768]
 [-0.1800261  -1.39045454 -0.44624945]
 [ 0.57028907  0.59333071  0.34512769]
 [-0.7775984   0.22934111 -0.40278742]
 [ 1.60610652  1.22962867  0.8132922 ]
 [-0.9100784  -0.8318138  -1.63211402]
 [ 0.71331896 -1.33662504 -0.2003138 ]
 [-1.02726706  1.3925347  -0.07097634]
 [ 0.20925473  0.69623403 -1.27713545]
 [ 0.3499596   0.15492091 -1.04838732]
 [ 1.50741045  0.36016048  0.71657877]
 [ 1.29405142  0.44878938 -0.7161128 ]
 [-1.3563561  -0.13475254 -0.99318601]
 [-0.2926833   1.34912764 -0.61305791]
 [-1.29960571 -0.48329572  1.4776761 ]
 [-0.78645463  0.52993357 -1.77537632]
 [-0.51136166 -0.66242918 -1.18623745]
 [ 0.10943817 -0.03636376  0.70814776]
 [-0.79575673 -0.87334131 -1.1731895 ]
 [-0.96897315  0.21841775

## Level 4: Boolean Masking & Fancy Indexing
Goal: Filter data without if statements.

Exercise 4.1: The Filter
Setup: arr = np.random.randint(0, 100, 20)
Task:

In [41]:
arr = np.random.randint(0, 100, 20)

1. Extract all odd numbers.

In [42]:
odd_indexes = (arr % 2 == 1) 
print(arr[odd_indexes])

[11 39 73 43 11 45  1]


2. Extract all numbers greater than 50. 

In [ ]:
greater_than_50_indexes=(arr > 50)
print(arr[greater_than_50_indexes])

[60 66 64 73 94]


3. Replace all numbers less than 0 with 0 (Create an array with negatives first to test). 

In [44]:
arr[arr < 0] = 0
print(arr)

[34 32 32 60 50 42 11 66 64 32 39 73 42 43 28 12 11 94 45  1]


4. Replace all odd numbers with -1. 

In [45]:
arr[arr % 2 == 1] = -1
print(arr)

[34 32 32 60 50 42 -1 66 64 32 -1 -1 42 -1 28 12 -1 94 -1 -1]


Exercise 4.2: The Complex Query
Setup: ages = np.random.randint(18, 90, 100) and income = np.random.randint(20000, 150000, 100).
Task:

In [46]:

ages = np.random.randint(18, 90, 100)
income = np.random.randint(20000, 150000, 100)

1. Find the average income of people older than 50.

In [47]:
mask_over_50 = ages > 50
average_income_over_50 = income[mask_over_50].mean()
print("Average income of people older than 50:", average_income_over_50)

Average income of people older than 50: 85935.0


2. Find the count of people who are younger than 30 AND have an income greater than 100,000.

In [48]:
mask_young_and_rich = (ages < 30) & (income > 100000)
print(mask_young_and_rich)
counter = np.sum(mask_young_and_rich)
print("Count of people younger than 30 with income > 100,000:", counter)

[False  True False False  True False False False False False False  True
 False False False False False False False False  True False False False
 False False False False False False False False False False  True False
 False False False False False False False False False False False False
 False False False False False  True False False  True False False False
 False False False False False False False False  True False False False
 False False False False False False False False False False False False
 False False False False False False False False False False False False
 False False False False]
Count of people younger than 30 with income > 100,000: 8


# Bonus:

## Level 5: Expert/Real-World Challenges
Goal: Solve complex logic puzzles using only vectorization.

Challenge 5.1: The One-Hot Encoder
Problem: Machine Learning models cannot understand words like "Red", "Green", "Blue". We convert them to binary vectors.
Input: labels = np.array([1, 0, 3, 2, 0, 1]) (Assume 4 categories: 0-3).
Task: Create a matrix of shape (6, 4) where the column corresponding to the label is 1, and others are 0.

In [53]:
colors = np.random.choice(["Red", "Green", "Blue"], size=20)
print(colors)
colors[colors=="Red"]=0
colors[colors=="Green"]=1
colors[colors=="Blue"]=2
print(colors)

['Green' 'Green' 'Green' 'Green' 'Green' 'Blue' 'Blue' 'Blue' 'Blue'
 'Green' 'Red' 'Green' 'Red' 'Blue' 'Green' 'Blue' 'Green' 'Blue' 'Red'
 'Red']
['1' '1' '1' '1' '1' '2' '2' '2' '2' '1' '0' '1' '0' '2' '1' '2' '1' '2'
 '0' '0']


## Challenge 5.2: Image Manipulation (The "Photoshop" Filter)
Problem: Images are just 3D arrays (Height, Width, RGB Channels).
Setup: Create a random "image" img = np.random.randint(0, 255, (100, 100, 3)).
Task:
    1. Grayscale conversion: Calculate the average of the 3 channels to create a 2D array (100, 100).
    2. Inversion: Create a negative of the image (New Value = 255 - Old Value).
    3. Masking: Set all pixels where the Red channel is > 150 to pure Black [0, 0, 0].

## Challenge 5.3: Distance Matrix (Euclidean)
Problem: You have 10 points in a 2D space. Calculate the distance between every pair of points.
Input: points = np.random.random((10, 2))
Task: Compute a (10, 10) matrix D where D[i, j] is the distance between point i and point j.
Hint: Use broadcasting with None (or np.newaxis).


## Challenge 5.4: Moving Average (Time Series Smoothing)
Input: A noisy stock price array prices = np.random.random(100).
Task: Calculate the 5-day moving average.
    • Element i of the result should be the mean of prices[i-4:i+1].
    • Constraint: You might use np.convolve for the cleanest solution, or advanced slicing.